File: mobile/iosApp/iosApp/Helpers/HtmlUtils.swift
import Foundation

/// Lightweight HTML stripping and entity decoding for plain-text contexts.
/// Full HTML rendering uses WKWebView in the detail screen.
extension String {

    /// Decode HTML character entities (&amp; &lt; &#39; &#x27; etc.).
    func strippingHTMLEntities() -> String {
        var result = self
        // Named entities
        let namedEntities: [String: String] = [
            "&amp;": "&", "&lt;": "<", "&gt;": ">",
            "&quot;": "\"", "&apos;": "'", "&nbsp;": "\u{00A0}",
            "&ndash;": "\u{2013}", "&mdash;": "\u{2014}",
            "&laquo;": "\u{00AB}", "&raquo;": "\u{00BB}",
            "&bull;": "\u{2022}", "&hellip;": "\u{2026}",
            "&copy;": "\u{00A9}", "&reg;": "\u{00AE}",
            "&trade;": "\u{2122}", "&euro;": "\u{20AC}",
        ]
        for (entity, char) in namedEntities {
            result = result.replacingOccurrences(of: entity, with: char)
        }
        // Decimal numeric entities: &#123;
        if let regex = try? NSRegularExpression(pattern: "&#(\\d+);") {
            let range = NSRange(result.startIndex..., in: result)
            let matches = regex.matches(in: result, range: range).reversed()
            for match in matches {
                if let codeRange = Range(match.range(at: 1), in: result),
                   let codePoint = UInt32(result[codeRange]),
                   let scalar = Unicode.Scalar(codePoint) {
                    let fullRange = Range(match.range, in: result)!
                    result.replaceSubrange(fullRange, with: String(Character(scalar)))
                }
            }
        }
        // Hex numeric entities: &#x1F4A9;
        if let regex = try? NSRegularExpression(pattern: "&#x([0-9a-fA-F]+);") {
            let range = NSRange(result.startIndex..., in: result)
            let matches = regex.matches(in: result, range: range).reversed()
            for match in matches {
                if let codeRange = Range(match.range(at: 1), in: result),
                   let codePoint = UInt32(result[codeRange], radix: 16),
                   let scalar = Unicode.Scalar(codePoint) {
                    let fullRange = Range(match.range, in: result)!
                    result.replaceSubrange(fullRange, with: String(Character(scalar)))
                }
            }
        }
        return result
    }

    /// Strip all HTML tags and decode entities, returning plain text.
    /// Block-level tags are replaced with newlines.
    func strippingHTML() -> String {
        guard !isEmpty else { return "" }
        var text = self
        // Replace <br> with newline
        text = text.replacingOccurrences(
            of: "<br\\s*/?>",
            with: "\n",
            options: [.regularExpression, .caseInsensitive]
        )
        // Replace block-level closing/opening tags with newline
        text = text.replacingOccurrences(
            of: "</?(p|div|li|tr|h[1-6])[^>]*>",
            with: "\n",
            options: [.regularExpression, .caseInsensitive]
        )
        // Strip remaining tags
        text = text.replacingOccurrences(
            of: "<[^>]*>",
            with: "",
            options: .regularExpression
        )
        // Decode entities
        text = text.strippingHTMLEntities()
        // Collapse whitespace
        text = text.components(separatedBy: .newlines)
            .map { $0.trimmingCharacters(in: .whitespaces)
                      .replacingOccurrences(of: "\\s{2,}", with: " ", options: .regularExpression) }
            .joined(separator: "\n")
        // Collapse multiple blank lines
        text = text.replacingOccurrences(of: "\n{3,}", with: "\n\n", options: .regularExpression)
        return text.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

File: mobile/shared/src/commonMain/kotlin/app/myfaq/shared/domain/HtmlUtils.kt
package app.myfaq.shared.domain

/**
 * Lightweight, pure-Kotlin HTML utilities for KMP.
 *
 * Used to strip tags and decode entities when displaying FAQ content
 * as plain text (list rows, search results, previews).
 * Full HTML rendering is handled by platform WebViews in the detail screen.
 */
object HtmlUtils {
    // Common HTML character-entity references (named).
    private val namedEntities =
        mapOf(
            "amp" to "&",
            "lt" to "<",
            "gt" to ">",
            "quot" to "\"",
            "apos" to "'",
            "nbsp" to "\u00A0",
            "ndash" to "\u2013",
            "mdash" to "\u2014",
            "laquo" to "\u00AB",
            "raquo" to "\u00BB",
            "bull" to "\u2022",
            "hellip" to "\u2026",
            "copy" to "\u00A9",
            "reg" to "\u00AE",
            "trade" to "\u2122",
            "euro" to "\u20AC",
            "pound" to "\u00A3",
            "yen" to "\u00A5",
            "deg" to "\u00B0",
            "times" to "\u00D7",
            "divide" to "\u00F7",
            "micro" to "\u00B5",
            "para" to "\u00B6",
            "sect" to "\u00A7",
            "sup1" to "\u00B9",
            "sup2" to "\u00B2",
            "sup3" to "\u00B3",
            "frac14" to "\u00BC",
            "frac12" to "\u00BD",
            "frac34" to "\u00BE",
        )

    private val tagRegex = Regex("<[^>]*>")
    private val entityRegex = Regex("&(#x?[0-9a-fA-F]+|[a-zA-Z]+);")
    private val whitespaceRegex = Regex("\\s{2,}")

    /**
     * Strip all HTML tags and decode entities, returning clean plain text.
     * Block-level tags (`<p>`, `<br>`, `<div>`, `<li>`) are replaced with
     * a single newline so the output preserves basic paragraph structure.
     */
    fun stripHtml(html: String): String {
        if (html.isBlank()) return ""

        var text = html

        // Replace block-level tags with newline before stripping
        text = text.replace(Regex("<br\\s*/?>", RegexOption.IGNORE_CASE), "\n")
        text = text.replace(Regex("</?(p|div|li|tr|h[1-6])[^>]*>", RegexOption.IGNORE_CASE), "\n")

        // Strip remaining tags
        text = tagRegex.replace(text, "")

        // Decode entities
        text = decodeEntities(text)

        // Collapse whitespace within lines but keep newlines
        text =
            text
                .lines()
                .joinToString("\n") { line -> whitespaceRegex.replace(line.trim(), " ") }

        // Collapse multiple blank lines
        text = text.replace(Regex("\n{3,}"), "\n\n")

        return text.trim()
    }

    /**
     * Decode HTML character entities (named, decimal, hex) in the given string.
     * Useful for cleaning question titles that contain encoded characters.
     */
    fun decodeEntities(text: String): String =
        entityRegex.replace(text) { match ->
            val entity = match.groupValues[1]
            when {
                entity.startsWith("#x", ignoreCase = true) -> {
                    val codePoint = entity.substring(2).toIntOrNull(16)
                    if (codePoint != null) codePoint.toChar().toString() else match.value
                }
                entity.startsWith("#") -> {
                    val codePoint = entity.substring(1).toIntOrNull()
                    if (codePoint != null) codePoint.toChar().toString() else match.value
                }
                else -> namedEntities[entity.lowercase()] ?: match.value
            }
        }

    /**
     * Return a short plain-text preview from HTML content,
     * truncated to [maxLength] characters with ellipsis.
     */
    fun preview(
        html: String,
        maxLength: Int = 150,
    ): String {
        val plain = stripHtml(html).replace('\n', ' ').trim()
        return if (plain.length <= maxLength) {
            plain
        } else {
            plain.take(maxLength).trimEnd() + "…"
        }
    }
}
